File: CastIt.Android/lib/models/dtos/responses/file_item_options_response_dto.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'package:json_annotation/json_annotation.dart';

part 'file_item_options_response_dto.freezed.dart';
part 'file_item_options_response_dto.g.dart';

@freezed
class FileItemOptionsResponseDto with _$FileItemOptionsResponseDto {
  factory FileItemOptionsResponseDto({
    required int id,
    required bool isVideo,
    required bool isAudio,
    required bool isSubTitle,
    required bool isQuality,
    required String text,
    required bool isSelected,
    required bool isEnabled,
    String? path,
  }) = _FileItemOptionsResponseDto;

  factory FileItemOptionsResponseDto.fromJson(Map<String, dynamic> json) => _$FileItemOptionsResponseDtoFromJson(json);

  // static List<String> get jsonKeys => [
  //       'Id',
  //       'IsVideo',
  //       'IsAudio',
  //       'IsSubTitle',
  //       'IsQuality',
  //       'Path',
  //       'Text',
  //       'IsSelected',
  //       'IsEnabled',
  //     ];
}

File: CastIt.Domain/Dtos/Responses/FileItemOptionsResponseDto.cs
namespace CastIt.Domain.Dtos.Responses
{
    public class FileItemOptionsResponseDto
    {
        public int Id { get; set; }
        public bool IsVideo { get; set; }
        public bool IsAudio { get; set; }
        public bool IsSubTitle { get; set; }
        public bool IsQuality { get; set; }
        public string Path { get; set; }
        public string Text { get; set; }
        public bool IsSelected { get; set; }
        public bool IsEnabled { get; set; }


        public static FileItemOptionsResponseDto ForVideo(int id, bool isSelected, bool isEnabled, string text)
        {
            return new FileItemOptionsResponseDto
            {
                Id = id,
                IsSelected = isSelected,
                IsEnabled = isEnabled,
                IsVideo = true,
                Text = text
            };
        }

        public static FileItemOptionsResponseDto ForAudio(int id, bool isSelected, bool isEnabled, string text)
        {
            return new FileItemOptionsResponseDto
            {
                Id = id,
                IsSelected = isSelected,
                IsEnabled = isEnabled,
                IsAudio = true,
                Text = text
            };
        }

        public static FileItemOptionsResponseDto ForLocalSubtitles(int id, string text, string path)
        {
            return new FileItemOptionsResponseDto
            {
                Id = id,
                IsSelected = true,
                IsEnabled = true,
                IsSubTitle = true,
                Text = text,
                Path = path
            };
        }

        public static FileItemOptionsResponseDto ForEmbeddedSubtitles(int id, bool isSelected, bool isEnabled, string text)
        {
            return new FileItemOptionsResponseDto
            {
                Id = id,
                IsSelected = isSelected,
                IsEnabled = isEnabled,
                IsSubTitle = true,
                Text = text
            };
        }
    }
}
